caikit/runtime/model_management/model_manager.py

```diff
@@ -450,6 +452,107 @@ def retrieve_model(self, model_id: str) -> ModuleBase:
+        )
+        try:
+            # If the model directory already exists, it's an error
+            model_dir = os.path.join(self._local_models_dir, model_id)
+            if os.path.exists(model_dir):
+                msg = f"Model '{model_id}' already exists"
+                raise CaikitRuntimeException(
+                    StatusCode.ALREADY_EXISTS, msg, {"model_id": model_id}
+                )
+
+            # Create the model directory directory
+            os.makedirs(model_dir)
+
+            # Write out all of the files
+            for fname, data in model_files.items():
+                fname = fname.strip()
+                if not fname:
+                    raise CaikitRuntimeException(
+                        StatusCode.INVALID_ARGUMENT,
+                        f"Got whitespace-only model file name: [{fname}]",
+                        {"model_id": model_id},
+                    )
+                fpath = os.path.join(model_dir, fname)
+                if not os.path.commonpath([model_dir, fpath]).lstrip(os.sep):
+                    raise CaikitRuntimeException(
+                        StatusCode.INVALID_ARGUMENT,
+                        f"Cannot use absolute paths for model files: {fname}",
+                        {"model_id": model_id},
+                    )
+
+                # Make sure intermediate dirs exist
+                parent_dir = os.path.dirname(fpath)
+                if os.path.relpath(parent_dir, model_dir) != ".":
+                    os.makedirs(parent_dir, exist_ok=True)
+
+                log.debug2(
+                    "Writing model file %s of size %s to %s", fname, len(data), fpath
+                )
+                with open(fpath, "wb") as handle:
+                    handle.write(data)
+
+            # Load the model
+            return self.load_model(
+                model_id=model_id,
+                local_model_path=model_dir,
+                model_type=self._LOCAL_MODEL_TYPE,
+                **kwargs,
+            )
+
+        except PermissionError as err:
+            raise CaikitRuntimeException(
+                StatusCode.FAILED_PRECONDITION,
+                f"Unable to save model (PermissionError): {err}",
+                {"model_id": model_id},
+            ) from err
+
+        except OSError as err:
+            raise CaikitRuntimeException(
+                StatusCode.UNKNOWN,
+                f"Unable to save model (OSError): {err}",
+                {"model_id": model_id},
+            ) from err
+
+    def undeploy_model(self, model_id: str):
+        """Remove the given model from the loaded model map and delete the
+        artifacts from the local models dir.
+        """
+        error.value_check(
+            "<RUN05068606E>",
+            self._local_models_dir,
+            "runtime.local_models_dir must be a valid path to undeploy models directly.",
+        )
+
+        # Get a handle to the LoadedModel
+        model = self.loaded_models.get(model_id)
```

Comment:
In case when we have multiple caikit pods, and only 1 of them have this model in memory, because we don't have a global shared state of models among caikit pods, this will not be able to "undeploy" the model right?

Reply:
These APIs will _only_ work if we have a shared volume for `local_models_dir`. That said, I think you found a race where the other replica has not yet loaded the deployed model (no inference, periodic poll hasn't triggered yet). In this case, the undeploy would fail. I'll fix that!